vm/src/gas/price_list.rs

```diff
@@ -0,0 +1,192 @@
+    /// in the runtime VM context.
+    pub ipld_get_base: i64,
+    pub ipld_get_per_byte: i64,
+
+    /// Gas cost (Base + len*PerByte) for any Put operation to the IPLD store
+    /// in the runtime VM context.
+    /// Note: these costs should be significantly higher than the costs for Get
+    /// operations, since they reflect not only serialization/deserialization
+    /// but also persistent storage of chain data.
+    pub ipld_put_base: i64,
+    pub ipld_put_per_byte: i64,
+
+    /// Gas cost for creating a new actor (via InitActor's Exec method).
+    /// Note: this costs assume that the extra will be partially or totally refunded while
+    /// the base is covering for the put.
+    pub create_actor_base: i64,
+    pub create_actor_extra: i64,
+
+    /// Gas cost for deleting an actor.
+    /// Note: this partially refunds the create cost to incentivise the deletion of the actors.
+    pub delete_actor: i64,
+
+    pub hashing_base: i64,
+    pub hashing_per_byte: i64,
+
+    pub compute_unsealed_sector_cid_base: i64,
+    pub verify_seal_base: i64,
+    pub verify_post_base: i64,
+    pub verify_consensus_fault: i64,
+}
+
+impl PriceList {
+    /// Returns the gas required for storing a message of a given size in the chain.
+    #[inline]
+    pub fn on_chain_message(&self, msg_size: i64) -> i64 {
+        self.on_chain_message_base + self.on_chain_message_per_byte * msg_size
+    }
+    /// Returns the gas required for storing the response of a message in the chain.
+    #[inline]
+    pub fn on_chain_return_value(&self, data_size: usize) -> i64 {
+        data_size as i64 * self.on_chain_return_value_per_byte
+    }
+    /// Returns the gas required when invoking a method.
+    #[inline]
+    pub fn on_method_invocation(&self, value: &TokenAmount, method_num: MethodNum) -> i64 {
+        let mut ret = self.send_base;
+        if value != &TokenAmount::zero() {
+            ret += self.send_transfer_funds;
+        }
+        if method_num != METHOD_SEND {
+            ret += self.send_invoke_method;
+        }
+        ret
+    }
+    /// Returns the gas required for storing an object
+    #[inline]
+    pub fn on_ipld_get(&self, data_size: usize) -> i64 {
+        self.ipld_get_base + data_size as i64 * self.ipld_get_per_byte
+    }
+    /// Returns the gas required for storing an object
+    #[inline]
+    pub fn on_ipld_put(&self, data_size: usize) -> i64 {
+        self.ipld_put_base + data_size as i64 * self.ipld_put_per_byte
+    }
+    /// Returns the gas required for creating an actor
+    #[inline]
+    pub fn on_create_actor(&self) -> i64 {
+        self.create_actor_base + self.create_actor_extra
+    }
+    /// Returns the gas required for deleting an actor
+    #[inline]
+    pub fn on_delete_actor(&self) -> i64 {
+        self.delete_actor
+    }
+    /// Returns gas required for signature verification
+    #[inline]
+    pub fn on_verify_signature(&self, sig_type: SignatureType, plain_text_size: usize) -> i64 {
```

Comment:
I notice in the lotus implementation that they are passing back a `SysErrInternal` exit code if the retrieval of the `costFn` fails. Do we need to match this exit code?

Also, I would include the following TODO as mentioned [here](https://github.com/filecoin-project/lotus/blob/testnet/3/chain/vm/gas.go#L59).

`this cost is not persistable, create a LinearCost{a,b} struct that has a .Cost(x) -> ax + b`

Reply:
nah, not needed, they are using a map to retrieve the function, we are just matching on all signature types, impossible for us to hit this